pipeline/data_store.py

```diff
@@ -0,0 +1,773 @@
+        Create a DataStore object from the given arguments.
+        """
+        if len(args) == 0:
+            raise ValueError('No arguments given to DataStore constructor!')
+        if len(args) == 1 and isinstance(args[0], DataStore):
+            return args[0]
+        else:
+            return DataStore(*args, **kwargs)
+
+    def __init__(self, *args, **kwargs):
+        """
+        See the parse_args method for details on how to initialize this object.
+        """
+        # these are data products that can be cached in the store
+        self.exposure = None  # single image, entire focal plane
+        self.image = None  # single image from one CCD
+        self.sources = None  # extracted sources (a SourceList object, basically a catalog)
+        self.wcs = None  # astrometric solution
+        self.zp = None  # photometric calibration
+        self.ref_image = None  # to be used to make subtractions
+        self.sub_image = None  # subtracted image
+        self.detections = None  # a SourceList object for sources detected in the subtraction image
+        self.cutouts = None  # cutouts around sources
+        self.measurements = None  # photometry and other measurements for each source
+
+        self.upstream_provs = None  # provenances to override the upstreams if no upstream objects exist
+
+        self.session = None  # optional database session
+
+        # these are identifiers used to find the data products in the database
+        self.exp_id = None  # use this and ccd_id to find the raw image
+        self.ccd_id = None  # use this and exp_id to find the raw image
+        self.im_id = None  # use this to specify an image already in the database
+
+        self.parse_args(*args, **kwargs)
+
+    def parse_args(self, *args, **kwargs):
+        """
+        Parse the arguments to the DataStore constructor.
+        Can initialize based on exposure and CCD ids,
+        or give a specific image id or coadd id.
+
+        Parameters
+        ----------
+        args: list
+            A list of arguments to parse.
+            Possible argument combinations are:
+            - exp_id, ccd_id: give two integers
+            - im_id: give a single integer
+
+        kwargs: dict
+            A dictionary of keyword arguments to parse.
+            Using named arguments allows the user to
+            explicitly assign the values to the correct
+            attributes. These are parsed after the args
+            list and can override it!
+
+        """
+        if len(args) == 0:
+            raise ValueError('Must provide at least one argument to DataStore constructor.')
+
+        if len(args) == 1 and isinstance(args[0], DataStore):
+            # if the only argument is a DataStore, copy it
+            self.__dict__ = args[0].__dict__.copy()
+            return
+
+        # catch any sessions given to the args list
+        sessions = [arg for arg in args if isinstance(arg, (sa.orm.session.Session, SmartSession))]
```

Comment:
This interface could potentially lead to trouble.  I can only see two ways of it being used, both of which lead to trouble.
(1) the datastore is initialized with a SmartSession inside a with block.  Once that with block ends, the SmartSession is no longer valid, but we'll probably be keeping the datastore around.
(2) the datastore istelf initializes the session, or is passed one outside a with block, and now we've got a session that's held open as long as the datastore exist, which goes against the "don't hold open sessions" issue we talked about at length before.

I think that datastore shouldn't keep a session inside itself.

Reply:
The reason for this design (which is not so clean), is that initialization goes through the DataStore, both for the individual objects (like Subtractor) or for the top level Pipeline object. I want to allow an external caller to provide a session if they choose, just like we give that option to all other function calls. 
The only way I could figure out how to do that was by parsing the session argument along with the other inputs. 
If you give a session externally, it remains in the DataStore. I'm not sure that's a problem. Lets go through examples:
```
with SmartSession() as session:
  ds = pipeline.run(exp_id, ccd_id, session=session)
  # session is in the ds, but the block is done here anyway
```

If you want to use individual objects you could conceivably do something weird like this:

```
with SmartSession() as session:
  ds = subtractor.run(im_id, session=session)  # parse the arguments into the ds
  ds = detector.run(ds)  # keep using the same open session
  # session in the ds, but it is closed afterwards

# outside the block there's still a reference to the session in ds:
ds = measurer(ds)  # accidentally use the same session
```

The new SQLA sessions can be safely closed and when accessed again they just start up a new connection using "autobegin" but if you want, we can disable that so that if someone tries to open a session that was closed by accident, it would raise an exception.

Reply:
I'd rather disable autobegin, because that's the kind of thing that will lead to database connections exploding without us realizing that it's happening.    Using that feature completely subverts the whole practice of controlling database connections using "with" blocks.

Indeed, I'd much rather not save the database connection in the object's self at all.  The only reason to do so is if we're going to use auto-re-opening things, which I think we shouldn't use.  Given that, it doesn't make any sense to save the session in self, because it will never be usable.

Reply:
OK I've added a workaround. 
The session is no longer saved in the data store. 
I suggest we turn off the session auto-begin only if we see there are too many open sessions.